src/compiler/checker.ts

```diff
@@ -15835,7 +15835,7 @@ namespace ts {
             const callSignatures = getSignaturesOfType(expressionType, SignatureKind.Call);
             if (callSignatures.length) {
                 const signature = resolveCall(node, callSignatures, candidatesOutArray);
-                if (getReturnTypeOfSignature(signature) !== voidType) {
+                if (!isInJavaScriptFile(signature.declaration) && getReturnTypeOfSignature(signature) !== voidType) {

```

Comment:
should not we check `functionSymbol.members` here to make sure the function has a `this.prop = ..` somewhere, and avoid something like `new getNumber()`